feat: network transaction state rework

Mirko-von-Leipzig opened the pull request:
This PR creates a new `State` tracker for the network transaction builder.

It is heavily inspired by the existing `PendingNotes` with some minor tweaks but with support for network account deltas and the ability to correlate all of this with inflight transactions.

I've also included a mockup in `NetworkTransactionBuilder::serve_new` of what the overall flow might look like - though I've tried to keep it fairly bare bones to keep the focus on the state portion. However I do recommend starting here as the entry-point since it might help inform the choices I've made.

Once again, I've left existing code in place as is.

### Architectural Overview

State tracks all current inflight txs and their impact on network account state.

State is updated purely via mempool events. The only exception is that an account is locked while it has a network tx in-progress _locally_.

The main loop consists of handling three possible events:

1. Mempool event received -> update state,
2. Interval tick -> select network tx (locking the account) and spawn a task for it,
3. A network tx task completes -> unlock the account if err (ok delegates this responsibility to the mempool event to ensure the account state is correct)

Spawning network txs allows for easy concurrency though this isn't strictly required. Its just a nice side-effect of the in-progress txs not actually mutating the `State`.

### State overview

Account deltas are stored separately to notes, and this should be fairly straight-forward code-wise I think. Notes are a fair bit more complicated to handle - and also drive which accounts are available for selection.

A specific account is locked once its been selected. This locking period is only until the result of the transaction is known locally - the transaction does not have to be committed; only either failed locally, or acknowledged by the mempool via an event. This is to prevent re-using the same notes and incomplete account state which would otherwise be incorrect i.e. this is a  temporary lock. I'm happy to hear alternatives here.

### Todo's

Main thing missing is migrating and extending the `PendingNotes` tests to the new `State`.

I'm unhappy with how complicated the `Notes` state feels, but I think there isn't really a way around this.

A separate PR would color in the `rest of the owl` for creating network transactions, and correct the startup procedures.

bobbinth reviewed (commented):
Thank you! Looks good! And a very detailed explanation of the architecture is greatly appreciated.

This is not a full review yet, but I left some small questions/comments inline. At the high-level, I think the approach makes sense.

> I'm unhappy with how complicated the `Notes` state feels, but I think there isn't really a way around this.

I'm a bit concerned about this too. I think the complexity that is there is fine for now - but it will only get more complex as we try to incorporate additional information (e.g., how many times executing a note failed, should a note be executed now or after some block etc.).

I tried to think of an alternative approach, and maybe making the state more "account-centric" could help - though, I haven't thought it through far enough to know if it actually makes sense. A rough sketch could look something like this:

```rust
pub struct State {
    /// AccountState here is kind of like a "mini-state" that keeps track of
    /// everything that is account-related. This includes the "inbox" of
    /// network notes (with their consumption status), in-flight account
    /// updates and transactions, the latest account state etc.
    accounts: BTreeMap<AccountId, AccountState>,

    /// A lookup table to help processing of transaction commits and reverts.
    txs_to_accounts: BTreeMap<TransactionId, AccountId>,

    /// A lookup table to help processing of newly received transactions.
    notes_to_accounts: BTreeMap<Nullifier, AccountId>

    /// Keeps track of accounts for which we are building a transaction but
    /// the mempool does not have this transaction yet.
    in_progress: BTreeMap<AccountId>
}
```

Mirko-von-Leipzig commented on crates/ntx-builder/src/state/account.rs:
```diff
@@ -0,0 +1,97 @@
+        self.tx_update(tx, TxUpdate::Revert);
+    }
+
+    fn tx_update(&mut self, tx: &TransactionId, change: TxUpdate) {
+        let Some(account) = self.txs.remove(tx) else {
+            return;
+        };
+
+        let deltas = self.deltas.get_mut(&account).unwrap();
+        match change {
+            TxUpdate::Commit => deltas.pop_front(),
+            TxUpdate::Revert => deltas.pop_back(),
+        };
+        if deltas.is_empty() {
+            self.deltas.remove(&account);
+        }
+    }
```
> This assumes that transactions are reverted and committed in order, right? Or at least that the order doesn't matter?

The order shouldn't matter, it relies purely on counts. We could try make the ordering matter using commitments, but it doesn't really change much aside from having to think about what to do if the ordering is wrong.

> Basically, what happens if we have 2 in-flight transactions against some account and both of them are committed (or revered)?

Then we would expect two calls to this function, and both changes would be reverted or committed.

As an example, if we have three inflight transactions in the order `A B C`, and two revert (that would have to be `B C`), then even if `C` is submitted first we still end up with overall state `A` at the end. And similarly for committed we would have `C` remaining.

Mirko-von-Leipzig commented on crates/ntx-builder/src/state/note.rs:
```diff
@@ -0,0 +1,249 @@
+/// if these are committed or reverted.
+#[derive(Default)]
+pub struct Notes {
+    /// Viable candidate accounts available for selection.
+    ///
+    /// They are guaranteed to have network notes.
+    queue: VecDeque<NetworkAccountPrefix>,
+
+    /// Accounts that have been selected and not yet deselected.
+    in_progress: HashSet<NetworkAccountPrefix>,
```
I'll try out the structure you've suggested; I had something similar in my earlier experiments but found it clunky. Notably a single transaction can affect multiple multiple account notes/nullifiers, but only a single account state.

But maybe there's a nice way using an additional transaction map.

drahnr commented on crates/ntx-builder/src/builder/mod.rs:
```diff
@@ -94,6 +101,81 @@ pub struct NetworkTransactionBuilder {
+            .context("failed to subscribe to mempool events")?;
+
+        let mut state = crate::state::State::new(unconsumed.into_iter());
+        let mut interval = tokio::time::interval(self.ticker_interval);
```
https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Burst is the default behaviour if we miss a tick. Not sure that's intended.

Mirko-von-Leipzig replied:
I think its.. probably fine. We should probably just `skip` instead

This does make me realize though what we definitely want to limit the overall number of in-progress tasks by skipping if `tasks.len() > N`. Thanks!

Mirko-von-Leipzig commented on crates/ntx-builder/src/builder/mod.rs:
```diff
@@ -94,6 +101,81 @@ pub struct NetworkTransactionBuilder {
+
+        loop {
+            tokio::select! {
+                _next = interval.tick() => {
```
Thanks - I've used it in the past so I kinda just assumed :sweat_smile: Always good to confirm though.

Mirko-von-Leipzig commented on crates/ntx-builder/src/state/account.rs:
```diff
@@ -0,0 +1,97 @@
+
+/// Tracks network account deltas for all currently inflight transactions.
+#[derive(Default)]
+pub struct AccountDeltas {
+    deltas: HashMap<NetworkAccountPrefix, VecDeque<NetworkAccountUpdate>>,
+    txs: BTreeMap<TransactionId, NetworkAccountPrefix>,
+}
```
> Given that the account ID prefix is globally unique

It is? Why is this then not just the account ID? :sweat_smile: Hashing purposes?

But yes that would resolve it then.

Mirko-von-Leipzig commented:
I've refactored towards something like @bobbinth suggested https://github.com/0xMiden/miden-node/pull/1039#pullrequestreview-2977864079.

I think this has worked out much better :fire:

As a summary:
- each account gets its own state, including deltas and notes
- the top level is now responsible for managing selection strategy

I've also simplified queue management imo.

Mirko-von-Leipzig commented on crates/ntx-builder/src/builder/mod.rs:
```diff
@@ -94,6 +101,87 @@ pub struct NetworkTransactionBuilder {
+                    let candidate = inflight_idx.remove(&task_id).unwrap();
+
+                    match completed {
+                        // Nothing to do. State will be updated by the eventual mempool event.
+                        Ok((_, Ok(_))) => {},
```
The concept is that we keep the account locked until either:

- the tx build fails, or
- the tx's mempool event arrives

Right now this is sort of implicit because the account gets unlocked when it receives a delta update. And since the account gets locked when it is selected this means that update event _must_ be that transaction.

I'm still trying to come up with something better, but its eluding me. Essentially the problem is that there are two possible things we can use to unlock the account again. Either when we know locally that the even was a success, or when that tx's mempool event arrives. The issue is that those two things can occur in either order. Both can carry the same information.

I've chosen to just use the mempool event since we need that code in any case. Though it does feel weird to notify of failure using the local way, and of success in a.. indirect way.

We could also say that we submit both to the state, and it just ignores the 2nd one - logic is already there to ignore duplicate txs.

Mirko-von-Leipzig replied:
The mental model I've used is

```
select account -> lock -> submit tx -> Ok  -> mempool -> event -> unlock
                                    -> Err -> unlock
```

bobbinth commented on crates/ntx-builder/src/state/mod.rs:
```diff
@@ -0,0 +1,278 @@
+}
+
+/// The impact a transaction has on the state.
+#[derive(Default)]
+struct Impact {
```
nit: I'd maybe name this `TransactionImpact` to improve readability a bit.

bobbinth commented on crates/ntx-builder/src/state/mod.rs:
```diff
@@ -0,0 +1,278 @@
+
+    /// Removes the account from tracking under the assumption that it is empty.
+    fn remove_account(&mut self, prefix: NetworkAccountPrefix) {
+        // We don't need to prune the inflight transactions because if the account is empty, then it
+        // would have no inflight txs.
```
Should we add an `assert!` for this?

bobbinth commented on crates/ntx-builder/src/state/mod.rs:
```diff
@@ -0,0 +1,278 @@
+
+    /// Handles [`MempoolEvent::TransactionsReverted`] events.
+    fn revert_transaction(&mut self, tx: TransactionId) {
+        let Impact { account_delta, notes, nullifiers } =
+            self.inflight_txs.remove(&tx).unwrap_or_default();
```
Would the `or_default()` part be unreachable here? Basically, if we are reverting a transaction, isn't it guaranteed that is is in `self.inflight_txs`?

Mirko-von-Leipzig replied:
The mempool sends _all_ reverted transactions but we only store txs with network-related impact. A tx with no network account delta and which produces no network notes or network nullifiers would not be in `self.inflight_txs`.

Might be clearer if I rewrite this as an early return with a comment though.

```rust
// Ignore transactions that aren't tracked.
//
// We only track transactions that have an impact on network state.
let Some(Impact { account_delta, notes, nullifier }) = self.inflight_txs.remove(&tx) else {
    return;
};
```

bobbinth replied:
Makes sense! Yes, using early return would make things more clear here.

bobbinth commented on crates/ntx-builder/src/state/mod.rs:
```diff
@@ -0,0 +1,278 @@
+
+    /// Handles [`MempoolEvent::BlockCommitted`] events.
+    fn commit_transaction(&mut self, tx: TransactionId) {
+        let Impact { account_delta, notes: _, nullifiers } =
+            self.inflight_txs.remove(&tx).unwrap_or_default();
```
Same question as above: under what conditions would we not find an inflight tx for a committed transaction?

bobbinth commented on crates/ntx-builder/src/state/account.rs:
```diff
@@ -0,0 +1,172 @@
+/// Tracks the state of a network account and its notes.
+#[derive(Default)]
+pub struct AccountState {
+    /// Inflight account updates in chronological order.
+    deltas: VecDeque<NetworkAccountUpdate>,
```
Is the reason we are tracking account deltas rather than full account states (i.e., `Account` structs) is memory efficiency? If so, I wonder if we should care about it for now. We could probably assuming that we could hold tens of thousands of account states in memory and once we go beyond that, we'd probably need a different solution anyways.

So, basically, if changing this to `VecDeque<Account>` simplifies things, I think we could consider this as a viable alternative.

Mirko-von-Leipzig replied:
We don't have access to the base account at this moment without reaching out to the store.

Reaching out to the store is of course possible, but then you run into potential race-conditions between the event, mempool and store states. My initial plan to avoid making things overly complex here was to download all account state on startup, but we decided to avoid that for now.

My current plan is to have these delta's and when the ntx is being built, it first reaches out to the store for the latest committed account state. There is still a race-condition here, where the first couple of delta's _might_ be already part of this committed state so we need to filter for those using the commitments. Which also means we probably want to use `TxAccountUpdate` instead of `AccountDelta` but that can be another PR.

bobbinth replied:
My thinking was that the state could be loaded on startup in 2 steps:

- Step 1: get all network notes from the store. From these, identify a set of relevant network accounts (i.e., accounts that have outstanding network notes for them).
- Step 2: get the accounts identified in Step 1 from the store.

At this point, 3 things could affect the account states:

- An account is updated via a transaction created by NTB. This will be handled naturally.
- A new network account is created by some transaction. This too should be handled naturally.
- A note is received for an account which has not been previously loaded from the store.

The last case would require additional handling, but given that NTB is currently the only one who's allowed to update network accounts, this shouldn't have any race conditions.

bobbinth replied:
> And this might include re-loading the account if it was dropped locally once all state was committed.

Yes, and in the future we could also add some kind of "timeout" logic so that we don't drop the account immediately after all of its notes have been processed, but after some time.

> Semi-related question: what do we do with network notes that have no account (yet)?
> 
> Do we keep them around in-case the account gets created afterwards, or do we just drop them?

Hmmm - good question! We could keep them around, but this would have to be time-bounded - so, doesn't really solve things.

What we could do instead is when a new network account gets created (and NTB receives the corresponding transaction), NTB could make a request to the store to get all network notes associated with this account. There probably is some kind of a race condition here - but should be possible to work around it.

bobbinth reviewed (commented):
Thank you! Looks good! And the structure feels much cleaner now. This is not a full review, but I left some comments inline - mostly nits and questions.

drahnr commented on crates/ntx-builder/src/builder/mod.rs:
```diff
@@ -94,6 +101,87 @@ pub struct NetworkTransactionBuilder {
+        loop {
+            tokio::select! {
+                _next = interval.tick() => {
+                    if inflight.len() > MAX_IN_PROGRESS_TXS {
```
The inflight set is limited by `MAX_INPROGRESS_TXS`. In each of the m`MAX_NOTES_PER_TX` can exist. How do we ensure they're not invalidating one another?

Mirko-von-Leipzig replied:
This follows from two facts:

1. A network note is bound to a single network account. As in, the note specifies a single network account that is allowed to consume it.
2. The `State` itself locks a network account once it has been selected until it has completed.

So each note can only be consumed by a specific account, and no account can have more than one transaction in-progress at once.

Mirko-von-Leipzig commented on crates/ntx-builder/src/state/account.rs:
```diff
@@ -0,0 +1,172 @@
+    available_notes: BTreeMap<Nullifier, NetworkNote>,
+
+    /// Notes which have been consumed by transactions that are still inflight.
+    nullified_notes: BTreeMap<Nullifier, NetworkNote>,
```
`#[derive(Hash)]` was only [recently](https://github.com/0xMiden/miden-base/issues/1472) added so it wasn't an option before.

I've created issue #1064 for this.
